src/audio/pipeline.c

```diff
@@ -198,7 +202,7 @@ static int pipeline_comp_complete(struct comp_dev *current,
 	current->period = ppl_data->p->ipc_pipe.period;
 	current->priority = ppl_data->p->ipc_pipe.priority;
 
-	pipeline_for_each_comp(current, ctx, dir);
+	pipeline_for_each_comp(current, calling_buf, ctx, dir);
```

Comment:
sure.

Reply:
Sorry, after think it more, @lgirdwood I think it's better to keep this calling_buf outside of the ctx, as the all the other ctx members are constant during the whole path walking, while this calling_buf is keep changing.

So better not to store it in the ctx otherwise it will be lost as the deeper layer will overwrite and break the upper layer value.

Reply:
Ok, what's the usage of calling_buf. Is it the current or next or previous source/sink buffer in the pipeline ? It sounds like we need to rename to better reflect the usage.

Reply:
> Ok, what's the usage of calling_buf. Is it the current or next or previous source/sink buffer in the pipeline ? It sounds like we need to rename to better reflect the usage.

It is the buffer ahead of the 'current' component, it could be one of the current component's source/sink buffer in the pipeline, depends on what direction we are walking. the calling_buf is used to check and make sure we are not iterating to the path which we have done to avoid dead loop, for example, in below path, we are walking .params() in this way: b1-> comp 1 -> params negotiation with b2, b3(b1 is excluded here for it is the calling_buf) done --> b3:

b3-----|
          v
b1 -> comp 1 --> b3
          ^
b2-----|

Reply:
Ok, this is really then the "next_buf". To avoid re-walking the same components you should add a member to component or buffer that is a flag for "walked already"

Reply:
> Ok, this is really then the "next_buf". To avoid re-walking the same components you should add a member to component or buffer that is a flag for "walked already"

I think it is not the "next", it should be the "previous"? adding a flag could be a choice, but then we might need to manage state of this flag (e.g. we need to do one more iteration to reset all of them once the walking is done, we even need to do this several times in one .params() of the pipeline, as you may know that the pipeline_comp_hw_params() is doing one more iteration before the real pipeline_comp_hw_params() is called).

Reply:
DAPM uses a "walked" flag when traversing the graph. It's reset before every walk. Please do the same here, sinec the DAPM version also survives loops (and this PR wont)

Reply:
We can add this flag to buffer only (no need for component), but to reset before every walk(e.g. .params(), .prepare(), .trigger(), .reset(), .copy()), we need to add a full iteration (to all buffers belong to or connected to the pipeline under iterated) ahead respectively, is this what you expected @lgirdwood ?